pam_eye.c

```diff
@@ -7,45 +8,87 @@
 #include <security/pam_modules.h>
 
 #define USERAGENT "pam_eye/1.0"
-#define TIMEOUT   10L
+#define DEFAULT_TIMEOUT_MS 200
 
-PAM_EXTERN int pam_sm_setcred(
-    pam_handle_t *pamh, int flags, int argc, const char **argv
+#define UNUSED(x) (x)
+
+PAM_EXTERN int pam_sm_close_session(
+    pam_handle_t *pamh, int flags,int argc, const char **argv
 ) {
-    return PAM_SUCCESS ;
+    UNUSED(pamh);
+    UNUSED(flags);
+    UNUSED(argc);
+    UNUSED(argv);
+
+    return PAM_SUCCESS;
 }
 
-PAM_EXTERN int pam_sm_authenticate(
+PAM_EXTERN int pam_sm_open_session(
     pam_handle_t *pamh, int flags,int argc, const char **argv
 ) {
+    UNUSED(pamh);
+    UNUSED(flags);
+
     if (argc == 0) {
         return PAM_SUCCESS;
     }
 
-    char host[256];
-    strncpy(host, argv[0], 256);
+    int timeout = DEFAULT_TIMEOUT_MS;
+    if (argc >= 2) {
+        timeout = atoi(argv[1]);
+    }
+
+    int debug = 1;
+    if (argc == 3) {
+        if (strcmp(argv[2], "nodebug") == 0) {
+            debug = 0;
+        }
+    }
 
-    char hostname[256];
-    gethostname(hostname, 256);
+    char report_url[256];
+    strncpy(report_url, argv[0], sizeof(report_url) - 1);
+    report_url[sizeof(report_url) - 1] = 0;
+
+    char local_hostname[HOST_NAME_MAX + 1];
+    gethostname(local_hostname, sizeof(local_hostname) - 1);
+    local_hostname[sizeof(local_hostname) - 1] = 0;
+
+    char url[
+        sizeof("http://") +
+        sizeof(report_url) +
+        sizeof("/") +
+        sizeof(local_hostname) +
+        1
```

Comment:
It's for ending '\0' byte (actually, there two of them already in report_url and local_hostname, but to be consistend I've decided to leave it as is.